walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,92 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.List;
+import java.util.Stack;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> ret = new ArrayList<>();
+    if (request.getDuration() <= 0) {
+      return ret;
+    }
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return ret;
+    }
+    List<TimeRange> occupiedBlocks = recordOccupiedBlocks(events, request.getAttendees());
+    ret = complement(TimeRange.WHOLE_DAY, occupiedBlocks);
+    ret.removeIf(tr -> tr.duration() < request.getDuration());
+    return ret;
+  }
+
+  /* Return list of blocks occupied by attendees, unordered */
+  private List<TimeRange> recordOccupiedBlocks(Collection<Event> events, Collection<String> names) {
+    List<TimeRange> occupiedBlocks = new ArrayList<>();
+    for (Event event : events) {
+      if (isOccupied(event, names)) {
+        occupiedBlocks.add(event.getWhen());
+      }
+    }
+    return occupiedBlocks;
+  }
+
+  /* Return true if anyone in the targetNames is involved in the event */
+  private boolean isOccupied(Event event, Collection<String> targetNames) {
+    for (String targetName : targetNames) {
+      if (event.getAttendees().contains(targetName)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /* Return a list of the available TimeRanges */
+  private List<TimeRange> complement(TimeRange wholeDay, List<TimeRange> mergedTimeRanges) {
+    Stack<TimeRange> complStack = new Stack<>();
+    Stack<TimeRange> rawStack = new Stack<>();
+
+    // TimeRanges are added in the order that they are returned by mergedTimeRanges's iterator
+    // The earliest (by start time) TimeRange is on top.
+    mergedTimeRanges.sort(TimeRange.ORDER_BY_START.reversed());
+    rawStack.addAll(mergedTimeRanges);
+
+    complStack.push(wholeDay);
+    while (!rawStack.isEmpty()) {
+      TimeRange tr1 = complStack.pop();
+      TimeRange tr2 = rawStack.pop();
+      // tr2.end()<=tr1.end()
+      complHelper(tr1, tr2, complStack);
+    }
+    return complStack;
   }
+
+  /* Compute the difference tr1\tr2, and push the result to the stack */
+  private void complHelper(TimeRange tr1, TimeRange tr2, Stack<TimeRange> stack) {
+    // tr2.end()<=tr1.end() guaranteed
+    if (tr1 == tr2) {
```

Comment:
Equality of objects in java only compares if they are the exact same object in memory.  This would evaluate to false if two different physical objects have the same data inside. The way to compare objects is by doing `tr1.equals(tr2)` Equals method needs to be overriden (which it is in the TimeRange file).

Reply:
ooh I didn't know that thanks for pointing it out!